### ISSUE TITLE
Pulls in CUPS daemon during start delaying boot

### ISSUE THREAD

**Issue by paulmenzel** [OpenPrinting/cups-filters#330]
The service unit `cups-browsed.service` requires `cups.service`

https://github.com/OpenPrinting/cups-filters/blob/1dc90ce0f0b7c050ca0bf6f24e889a0eda19c1fb/utils/cups-browsed.service#L1-L11

causing the CUPS daemon (also when configured socket activatable) to be started during boot despite it only being needed when actually printing on most desktop systems.

```
$ systemd-analyze critical-chain
[…]
graphical.target @2.218s
└─multi-user.target @2.218s
  └─cups-browsed.service @2.218s
    └─cups.service @1.292s +924ms
      └─network.target @1.291s
        └─NetworkManager.service @1.185s +106ms
          └─dbus.service @1.182s
            └─basic.target @1.178s
              └─sockets.target @1.178s
                └─uuidd.socket @1.178s
                  └─sysinit.target @1.173s
                    └─systemd-backlight@backlight:intel_backlight.service @2.162s +5ms
                      └─system-systemd\x2dbacklight.slice @1.363s
                        └─system.slice @334ms
                          └─-.slice @334ms
```

It would be great, if cups-browsed was started on demand, when users want to print something or look at the printers.

**Comment by paulmenzel** [comment]
(Note, delay means, that it’s started in parallel during bootup, consuming resources like CPU causing other essential services to start longer. It does *not* mean, that the ordering causes a delay, that another target/service depends on `cups-browsed.service` to have started.)

**Comment by paulmenzel** [comment]
```
$ pgrep cups-browsed
$ lsof -p600
[…]
cups-brow 600 root    0r      CHR                1,3      0t0        4 /dev/null
cups-brow 600 root    1u     unix 0x00000000bfd0e2ba      0t0    18490 type=STREAM
cups-brow 600 root    2u     unix 0x00000000bfd0e2ba      0t0    18490 type=STREAM
cups-brow 600 root    3u     unix 0x0000000043ce89eb      0t0    15351 type=STREAM
cups-brow 600 root    4u     unix 0x000000000a79b37e      0t0    15357 type=STREAM
cups-brow 600 root    5u  a_inode               0,13        0     8983 [eventfd]
cups-brow 600 root    6u     unix 0x00000000fffa10bf      0t0    19475 type=STREAM
cups-brow 600 root    7u     IPv4              18632      0t0      UDP *:631 
cups-brow 600 root    8u  a_inode               0,13        0     8983 [eventfd]
cups-brow 600 root    9u     unix 0x00000000651b026f      0t0    18633 type=STREAM
cups-brow 600 root   10u  a_inode               0,13        0     8983 [eventfd]
```

**Comment by tillkamppeter** [comment]
TL; DR: This is a bug of the print dialogs, they need to use the current, destination-based CUPS APIs. Therefore I am closing this bug as it is not a bug in cups-filters.

cups-browsed in most cases serves for making print dialogs compatible with modern CUPS and less to manage how network printers get shared (clustering, ...).

Modern CUPS handles driverless IPP network printers and shared CUPS queues from remote servers with temporary queues, queues which only get created if one accesses them, for getting the queue's capabilities or for printing, and they get removed after one minute of inactivity.

Print dialogs, created many years ago, uses obsolete CUPS APIs. which do not support these temporary print queues. They only see the permanent CUPS queues, usually created manually.

In the beginning, CUPS used its own protocol for broadcasting the printer it shares and for browsing shared printers of other, remote CUPS daemons to be used locally. This made network printing easy. Simply make your local printer(s) shared and everyone in your local network sees them in print dialogs. These queues (the automatically created ones on the clients) were also temporary/virtual, but the APIs of that times and still used by current print dialogs exposed these printers.

Then a decade ago CUPS dropped this protocol in favor of the standard DNS-SD, but it only registered its shared printers with DNS-SD. It did not browse the shared printers of other, remote CUPS daemons for local use. This made me creating cups-browsed, to get a transparent transition for the users. cups-browsed was picking up DNS-SD-advertised printers, in the beginning remote CUPS queues, later also IPP network printers, and creating local queues pointing to them and removing them when the remote printer disappears or when cups-browsed gets stopped (so queues are pseudo-temporary, but actually permanent from the CUPS point of view, so no problem for print dialogs). This way everything worked as in good old CUPS broadcasting/browsing times and distro users did not perceive any change.

cups-browsed got standard in all Linux distribution and that time to overcome the changed of CUPS to server-side-only DNS-SD.

Later on a new change: CUPS got automatic pick-up of remote printers again, via the above-mentioned temporary queues. They are actually exposed, especially even before being created, via the new CUPS destination API. Even with this change in place for some years the print dialogs still use the old interface, so would not see the temporary queues. To work around this problem the distros simply keep cups-browsed in place as it auto-creates permanent queues (and has even configuration options to not badly interfere with CUPS' concept of temporary queues).

Both cups-browsed and CUPS support running the daemon on-demand. I have taken care of this especially during the time when Ubuntu Touch (for mobile devices like phones) was developed to have a resource-saving way to get printing support. Note that this was before the concept of temporary CUPS queues for IPP printers was introduced.  This was working the way that the print dialog when it gets opened fires up avahi-daemon and cups-browsed and cups-browsed in turn fires up CUPS, then it opens the list of available printers (queues created by cups-browsed), the user selects a printer, sets options, prints, and finally closes the dialog, now avahi-daemon and cups-browsed get shut down (or simply only avahi-daemon and cups-browsed closes by itself then when not seeing DNS-SD printers any more) and finally CUPS terminates when cups-browsed is not there any more.

As a conclusion, the place where we can fix is issue is exclusively in the print dialog. If we go the concept of Ubuntu Touch, we would need to add functionality to the print dialogs to start cups-browsed in the beginning and stop it in the end. But with the temporary queues for IPP printers in modern CUPS we do not need this any more, as I wrote before, cups-browsed is currently only running as workaround due to the print dialogs using obsolete CUPS APIs. So what we have to do is simple, **make the print dialogs use the current destination API of CUPS**, so that cups-browsed does not need to be used on ordinary desktop setups.

Closing, as this is not a bug in cups-filters ...

**Comment by paulmenzel** [comment]
Thank you for the detailed explanation.

[…]

> Both cups-browsed and CUPS support running the daemon on-demand. I have taken care of this especially during the time when Ubuntu Touch (for mobile devices like phones) was developed to have a resource-saving way to get printing support. Note that this was before the concept of temporary CUPS queues for IPP printers was introduced. This was working the way that the print dialog when it gets opened fires up avahi-daemon and cups-browsed and cups-browsed in turn fires up CUPS, then it opens the list of available printers (queues created by cups-browsed), the user selects a printer, sets options, prints, and finally closes the dialog, now avahi-daemon and cups-browsed get shut down (or simply only avahi-daemon and cups-browsed closes by itself then when not seeing DNS-SD printers any more) and finally CUPS terminates when cups-browsed is not there any more.

So, cups-browsed does not need to be running all the time and the service unit can be disabled?

**Comment by tillkamppeter** [comment]
It depends on the situation. If you manually create print queues for all the printer you want to use (works also with driverless printers, thanks to the `driverless` utility they appear in printer setup tools) you do not need cups-browsed to auto-create queues. Then cupsd can work perfectly on-demand. Check whether you also need to turn off the web interface of CUPS to assure that it does not stay in memory.
cups-browsed only serves for automatic setup and removal of local CUPS queues for both IPP printers and remote CUPS printers.
